### PR TITLE
[SPARK-7885][SQL]add config to control map aggregation in spark sql

### PR DESCRIPTION
**Pull request by jeanlyn**
[SPARK-7885](https://issues.apache.org/jira/browse/SPARK-7885),we add ~~`spark.sql.partialAggregation.enable`~~ `spark.sql.preAggregation.enable`,it's true by default,we can set false to make map aggregation unable to avoid gc problem.For example,we run the sql

``` sql
insert overwrite table groupbytest
select sale_ord_id as order_id,
      coalesce(sum(sku_offer_amount),0.0) as sku_offer_amount,
      coalesce(sum(suit_offer_amount),0.0) as suit_offer_amount,
      coalesce(sum(flash_gp_offer_amount),0.0) + coalesce(sum(gp_offer_amount),0.0) as gp_offer_amount,
      coalesce(sum(flash_gp_offer_amount),0.0) as flash_gp_offer_amount,
      coalesce(sum(full_minus_offer_amount),0.0) as full_rebate_offer_amount,
      0.0 as telecom_point_offer_amount,
      coalesce(sum(coupon_pay_amount),0.0) as dq_and_jq_pay_amount,
      coalesce(sum(jq_pay_amount),0.0) + coalesce(sum(pop_shop_jq_pay_amount),0.0) + coalesce(sum(lim_cate_jq_pay_amount),0.0) as jq_pay_amount,
      coalesce(sum(dq_pay_amount),0.0) + coalesce(sum(pop_shop_dq_pay_amount),0.0) + coalesce(sum(lim_cate_dq_pay_amount),0.0) as dq_pay_amount,
      coalesce(sum(gift_cps_pay_amount),0.0) as gift_cps_pay_amount ,
      coalesce(sum(mobile_red_packet_pay_amount),0.0) as mobile_red_packet_pay_amount,
      coalesce(sum(acct_bal_pay_amount),0.0) as acct_bal_pay_amount,
      coalesce(sum(jbean_pay_amount),0.0) as jbean_pay_amount,
      coalesce(sum(sku_rebate_amount),0.0) as sku_rebate_amount,
      coalesce(sum(yixun_point_pay_amount),0.0) as yixun_point_pay_amount,
      coalesce(sum(sku_freight_coupon_amount),0.0) as freight_coupon_amount
from        ord_at_det_di
where       ds = '2015-05-20'
group  by   sale_ord_id
```

use 6 executor, each executor has 8GB memory and 2 cpu,we got gc problems during the map aggregation and finally the executor crash
5869030a-d924-4249-9e1d-c637caa9363a 

When we set ~~`spark.sql.partialAggregation.enable`~~ `spark.sql.preAggregation.enable` false ,the sql run in 2 min

### CODE REVIEW

**Comment by JoshRosen** [comment]
I guess that the GC benefits here are coming from the fact that the map-side aggregation must, in the worst case, allocate a map entry for every possible key, whereas the reduce-side aggregation only needs to handle 1/N of the keys.  I could imagine this being helpful for cases where you have a large number of unique keys and a high degree of skew, since the unique keys won't benefit from pre-aggregation and will just increase the size of the map.

**Comment by JoshRosen** [comment]
One note on naming: I worry that "partial aggregation" will be confused with [partial pre-aggregation](http://ieeexplore.ieee.org/xpl/login.jsp?tp=&arnumber=994787&url=http%3A%2F%2Fieeexplore.ieee.org%2Fxpls%2Fabs_all.jsp%3Farnumber%3D994787), a different aggregation optimization:

> Partial preaggregation is a simple data reduction operator that can be applied to aggregation queries. Whenever we group and aggregate on a column set G, we can preaggregate on any column set that functionally determines G. Preaggregation can be used, for example, to reduce the input size to a join. Regular aggregation reduces the input to one record per group. Partial preaggregation exploits the fact that preaggregation need not be complete-if multiple records happen to be output for a group, they will be combined into the same group by the final aggregation. 

Instead, maybe it would be clearer to name the option "preAggregation" since it completely disables the map-side aggregation.  I guess that the current name is a carryover from the `PartialAggregation` class name.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 1]
```diff
@@ -119,6 +119,15 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
   object HashAggregation extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // We can cancel the map aggregation due to GC problems
+      case logical.Aggregate(groupingExpressions, aggregateExpressions, child)
+        if !sqlContext.conf.partAggregationEnabled =>
+        execution.Aggregate(
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: I think that the decision of whether to pre-aggregate or not is orthogonal to whether we should use a generated aggregate or the regular interpreted aggregate operator.  Therefore, if we decide to implement this feature then I think we'll have to refactor this block in order to support both operators.  Maybe there's a nice common interface / builder pattern that we could use to abstract over the actual aggregate operator that's used.

**Comment by JoshRosen** [comment]
Note: I'd wait for an actual SQL committer to comment on this before making any changes based on my comments, since I'll defer to them to make the call on whether to include this feature.  Just wanted to comment [since aggregation is a topic dear to my heart](https://github.com/JoshRosen/aggregation-optimization-paper).

**Comment by rxin** [comment]
@jeanlyn a better way to do this is to actually run some partial aggregation, and if we see no reduction in data size, stop the partial aggregation and just run the aggregate methods quickly. I think that might be somewhat tricky to do based on the current implementation of aggregates, but should be much easier once we refactor aggregate/UDAF interface for 1.5.

**Comment by jeanlyn** [comment]
Thanks @JoshRosen @rxin for comment.I had meet at lease two group by cases  of our production environment run long GC time and finally executor crash.These case has a comment characteristic: The input split is large(more than 128MB),and the key is also large.But had no data skew.because when i turn `spark.sql.partialAggregation.enable` false,we can get the result quickly.And each reduce task handle the data sizes almost the same.
@rxin Do you had more informations about `refactor aggregate/UDAF interface`.Thanks!

**Comment by rxin** [comment]
That patch is in this pull request: https://github.com/apache/spark/pull/5542

And jira ticket: https://issues.apache.org/jira/browse/SPARK-4233

**Comment by jeanlyn** [comment]
Thanks @rxin for informations.I close this PR for now and will reopen it once be optimized.

**Pull request closed without merging**